CI progress

kylecordes opened the pull request:
Starting E2E implementation, progress on #430 and closes #431.

This implementation **assumes Mitosis is intended to emit files ready to ship/consume as an NPM package**, perhaps by a sibling project in a monorepo. That wasn't written down anywhere, but I saw hints this direction.

See the commit comments for more details.

I first started with a single workspace to test multiple targets, but found I was creating a multi-project mechanism within a single workspace... which is Yarn's job. So I switched to one workspace per target.

Free payoff of multiple workspaces: the yarn foreach mechanism parallelizes e2e execution!

This PR has several of eventually N workspaces, one for each target, or occasionally (as with React) JS and TS variations, where both kinds of consumer code are common. This also serves as an example for users on how to have a monorepo, define components in a Mitosis project, and consume them from a project based on React/Vue/Solid/etc.

I have in mind Playwright to verify that the output runs, shows a correct GUI, responds to clicks, etc. Mitosis working **e2e** would mean a pile of components are created once, compiled to N targets, and all N targets pass the same web-automation test suite. (Plus later other consideration for non-web target ideas.)

To try the progress so far: `yarn ci`

I have no idea how close this is to what the Mitosis team has in mind for E2E, feel free to savage it, I will not be mortally offended.

samijaber commented on packages/e2e-vue3-ts/vite.config.ts:
```diff
@@ -0,0 +1,7 @@
+// https://vitejs.dev/config/
+export default defineConfig({
+  plugins: [vue()],
+});
```
PS: Our vue 3 output is actually broken in Mitosis, but that's something that's going to be fixed in the _very_ near future. So we might want a vue 2 server here instead for now, or just drop the vue 3 package.

samijaber replied:
oh wow...or it's not broken! 😄 This works for me locally 🎉

kylecordes replied:
The only output is under “nuxt” and appears to work with Vue 3.

It barely works - can't even do two components because the inter-component import doesn’t work.

But at least this will lock in this first increment of working. Enable the ratchet of progress.

kylecordes commented:
I believe this is ready to go, except for one problem:

**We treat the Vue output as Vue3**, and it works for this simple case. I've attempted to switch to Vue 2, but not been able to get all the wiring to work. Perhaps we can proceed with this as it is, and seek a Vue 2 expert to take a look?

Avenues of further work:

* Cover more targets; but some of them are waiting for mitosis fixes.
* Split the trivial e2e-app into 2 components - this need a Vue fix first.
* Store the E2E output reports from Playwright, in case a developer wants to look at them without running locally.
* Split apart the results in GitHub actions somehow, so that failure details are more easily spotted without studying lengthy build output.
* Consider a more powerful build tool, either with or instead of yarn workspaces; currently every build rebuilds everything, which is unnecessary.
* Consider CI caching or a CI container which has the needed browser pre-installed, to speed up CI.

samijaber commented on packages/e2e-solid/playwright.config.ts:
```diff
@@ -0,0 +1,6 @@
+import { configFor } from '@builder.io/e2e-app-spec/shared.config';
+
+export default configFor(
+  'yarn workspace @builder.io/e2e-solid run serve --port 7501',
+  7501,
```
I wonder if we can change `configFor`'s logic so that we don't have to include the `--port $PORT` in the command, and instead append it in `configFor`. That would remove the duplication we're seeing each time.

kylecordes replied:
Yes, there are def ways to make this better. Important to do now? Or followup?

samijaber replied:
Followup is totally fine!

samijaber commented on packages/e2e-solid/package.json:
```diff
@@ -0,0 +1,23 @@
+    "build": "vite build",
+    "dev": "vite",
+    "e2e": "playwright test",
+    "playwright-install": "playwright install --with-deps",
```
Is there a particular reason why this command has to exist within one of the `e2e` packages, and not in the root `package.json`?

kylecordes replied:
I put it here because this package already had the dependency. Because of a VERY long story around how Typescript looks up types, this project uses Yarn in a mode that doesn't consolidate/hoist, so this is slightly faster. Running it from one of the points where we execute Playwright seemed expedient, though a bit ugly.

samijaber replied:
Love a long story around Typescript & types 😄.

Not a big deal at all, let's do this in a follow up if at all

samijaber commented on packages/e2e-solid/package.json:
```diff
@@ -0,0 +1,23 @@
+    "solid-styled-components": "^0.28.1"
+  },
+  "devDependencies": {
+    "typescript": "^4",
```
One thing: could we add `playwright` as a devDependency to each `e2e` project as well? If we don't, it's impossible to run `yarn run e2e` within just one of the e2e packages, which I think will be needed when debugging individual outputs 🤔. What do you think? 

Can make it a very wide version range like you did with TS here, so that they all use the same version.

samijaber replied:
This is also ok in a followup PR @kylecordes. Let's merge this thing since it's all good and working!!

samijaber replied:
Oh, just saw your answer right before mine. Adding some docs would be good to clarify that, as people might try to do what I did (cd into a package, then run playwright there and see it fail).

samijaber reviewed (commented):
I am _so_ excited for this PR. Absolutely fantastic effort, thank you so much @kylecordes.

I have left a few small comments, but once they're addressed this thing is getting shipped ASAP.

samijaber commented on packages/e2e-app/src/components/my-component.lite.tsx:
```diff
@@ -0,0 +1,60 @@
+import '@builder.io/mitosis/dist/src/jsx-types';
```
(This is more for my own out-loud thoughts, nothing actionable in this PR!)

samijaber replied:
Yeah for sure, let's not creep scope this!

kylecordes commented on package.json:
```diff
@@ -13,18 +13,21 @@
+    "lint:prettier": "yarn run prettier --check",
     "prerelease": "yarn build",
-    "prettier": "prettier '{packages,docs}/**/*.{js,jsx,ts,tsx,json,md,html}'",
+    "prettier": "prettier --loglevel warn '{packages,docs}/**/*.{js,jsx,ts,tsx,json,md,html}'",
```
The CI output is noisy - this is a tiny step in a less noisy direction. Noone seems to know why prettier emits so much output when it takes no action this setting makes it not do that, many prettier users think this should be the default. I apologize for the scope creep toward "make the CI better generally"...

samijaber replied:
No worries! Was just curious about the impact. Fine by me, thanks for cleaning it up 👍🏽
